spec/requests/included_resources_spec.rb

```diff
@@ -7,20 +7,20 @@
   subject { last_response }
   let(:json_included) { JSON.parse(last_response.body)['included'] }
 
-  let(:comments_policy_scope) { Comment.none }
+  let(:comments_policy_scope) { Comment.all }
   let(:article_policy_scope) { Article.all }
   let(:user_policy_scope) { User.all }
 
   before do
-    allow_any_instance_of(ArticlePolicy::Scope).to receive(:resolve).and_return(
-      article_policy_scope
-    )
-    allow_any_instance_of(CommentPolicy::Scope).to receive(:resolve).and_return(
-      comments_policy_scope
-    )
-    allow_any_instance_of(UserPolicy::Scope).to receive(:resolve).and_return(
-      user_policy_scope
-    )
+    allow_any_instance_of(ArticlePolicy::Scope).to receive(:resolve) do |policy_scope|
+      article_policy_scope.merge(policy_scope.scope.all)
+    end
+    allow_any_instance_of(CommentPolicy::Scope).to receive(:resolve) do |policy_scope|
+      comments_policy_scope.merge(policy_scope.scope.all)
+    end
+    allow_any_instance_of(UserPolicy::Scope).to receive(:resolve) do |policy_scope|
+      user_policy_scope.merge(policy_scope.scope.all)
+    end
```

Comment:
I'm having a bit of a hard time figuring out what these policy scope changes do. It seems like it's overriding the policy scope stubbing to always return all results?

Maybe there could be some sort of a documentation comment here to tell why this is necessary.

Reply:
I wanted to get rid of these entirely, but that required a much larger overhaul of the tests than I wanted to take on.

As previously written, this spec never checked which `has_many` records were returned except in situations where the spec stubbed the `comments_policy_scope` to be `Comment.limit(1)`, and it masked a fundamental problem with the mock when I added additional scenarios to test.

https://github.com/venuu/jsonapi-authorization/blob/05785ff8c3d841eae2a959c59f80b9ad408c5214/lib/jsonapi/authorization/pundit_scoped_resource.rb#L15-L28
The value of `policy_scope!` is supposed to be the records that the user has access to, further restricted by the specific `record_or_records` in the association. A simple `and_return` like what was here before would return all records in the user's scope, even if it had nothing to do with the `@model` under test. For example, a user who has access to all articles and looked up the articles written by a single user would get a list of all articles instead of just the articles written by that user.

`merge` acts like a sql `AND`, so this change allows the stubbed scope to return something more similar to how it should actually behave outside of tests.

Will add a comment for this.

Reply:
Perfect, thanks for clarifying!